DOC: pd.read_csv doc-string clarification #11555

frankcleary opened the pull request:
closes #11555 
Updated IO Tools documentation for read_csv() and read_table() to be consistent with the doc-string.

jreback commented:
what exactly changed here?

jreback commented:
can you verify that these are in numpy-doc format, see [here](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

mostly it comes down to a parameter like:

```
parm : datatype, optional
     description
```

jreback commented:
```
Parameters

Description of the function arguments, keywords and their respective types.

Parameters
----------
x : type
    Description of parameter `x`.
Enclose variables in single backticks. The colon must be preceded by a space, or omitted if the type is absent.

For the parameter types, be as precise as possible. Below are a few examples of parameters and their types.

Parameters
----------
filename : str
copy : bool
dtype : data-type
iterable : iterable object
shape : int or tuple of int
files : list of str
If it is not necessary to specify a keyword argument, use optional:

x : int, optional
Optional keyword parameters have default values, which are displayed as part of the function signature. They can also be detailed in the description:

Description of parameter `x` (the default is -1, which implies summation
over all axes).
When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first:

order : {'C', 'F', 'A'}
    Description of `order`.
When two or more input parameters have exactly the same type, shape and description, they can be combined:

x1, x2 : array_like
    Input arrays, description of `x1`, `x2`.
```

frankcleary commented:
In terms of what was changed: I updated the IO Tools description of `read_csv()` that appears [here](http://pandas.pydata.org/pandas-docs/stable/io.html#csv-text-files) to match the doc string. I thought that's what the bug was getting at. I also slightly changed the wording describing the default behavior of the `header=` option.

I can go through and make any changes to put the doc string in numpy format, would you prefer that as a separate PR and/or should I drop the IO Tools doc changes?

jreback commented:
@frankcleary thanks. can you enumerate what changed (here), though. just for my edification.

yes, would like to conform to numpy-doc in this PR.

frankcleary commented:
Makes sense, I'll do that and update this PR later this week.

jorisvandenbossche commented:
@jreback The list is in io.rst, not in the docstring. So I think it does not need to follow the numpy-doc format?

jreback commented:
@jorisvandenbossche hmm, good point. I am not sure *what* we actually do as we don't have too many places where we go thru the args of functions. is there a consistent pattern now (anywhere)? any reason *not* to use a doc-string format?

frankcleary commented:
The only other place I could find in the docs with a similar list of args is for `concat` in [merging.rst](http://pandas.pydata.org/pandas-docs/stable/merging.html#concatenating-objects). This also appears to have lagged behind the [`concat` docstring](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) in terms of updates (for example there's no  `copy=...`). I wonder if it's better to remove these lists from the *.rst documentation in favor of links to the docs generated from the docstrings themselves?

jreback commented:
@frankcleary the ideal thing would be to render the doc-strings in-line. as the out-of-syncness is annoying. Though if you simply want to fix it we can make that another issue.

frankcleary commented:
Looks like sphinx can do that, updated. Here's a sample of what the result looks like in the io docs:


![screen shot 2015-12-13 at 9 39 32 pm](https://cloud.githubusercontent.com/assets/5619798/11773688/0c65c1fc-a1e2-11e5-95fe-1ca9571e9dfb.png)

jreback commented:
@jorisvandenbossche how's this look to you

jorisvandenbossche commented:
Hmm, personally, if it looks exactly the same as the generated API page, I would just link to it rather than including it. 
But I also understand that keeping a separate list of the parameters up to date is more work, but I think it is nicer for the tutorial docs.

jreback commented:
so after seeing this, I think I agree with @jorisvandenbossche 

let's revert back to just an updated list (and hopefully we can keep this updated).

@frankcleary can you update.

frankcleary commented:
Sure, I'll get to it this week.

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
-    'round_trip' for the round-trip converter.
+  - ``filepath_or_buffer``: str or file handle /
+    :class:`python:io.StringIO`. The string could be a URL. Valid URL schemes
+    include http, ftp, s3, and file. For file URLs, a host is expected. For
```
should we mention the LocalPath here as well? (its in the above)

frankcleary replied:
Fixed

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+  - ``sep``: str, defaults to ``','`` for
+    :func:`~pandas.io.parsers.read_csv`, ``\t`` for
+    :func:`~pandas.io.parsers.read_table`. Delimiter to use. If sep is ``None``,
+    will try to automatically determine this. Regular expressions are accepted.
```
mention that regexs force the python engine

frankcleary replied:
Fixed

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+    used to escape delimiter when quoting is ``QUOTE_NONE``.
+  - ``dtype``: Type name or dict of column -> type, default ``None``. Data type
+    for data or columns. E.g. ``{'a': np.float64, 'b': np.int32}`` (unsupported
+    with ``engine='python'``).
```
menton ``str`` or ``object`` to preserve and not interpret for dtypes

frankcleary replied:
Fixed

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+    multi-index on the columns e.g. ``[0,1,3]``. Intervening rows that are not
+    specified will be skipped (e.g. 2 in this example is skipped). Note that
+    this parameter ignores commented lines and empty lines if
+    ``skip_blank_lines=True``, so header=0 denotes the first line of data
```
needs a ``-``

frankcleary replied:
It's actually part of the paragraph for the header option

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+  - ``true_values``: list, default ``None``. Values to consider as ``True``.
+  - ``false_values``: list, default ``None``. Values to consider as ``False``.
+  - ``keep_default_na``: bool, default ``True``. If na_values are specified and
+    keep_default_na is ``False`` the default NaN values are overridden,
```
maybe add in the list of default NA values here? (or in ``na_values``)

frankcleary replied:
Fixed by inserting the actual values, although I'm not sure if the added complexity is worth it.

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+    specifies combining multiple columns then keep the original columns.
+  - ``date_parser``: function, default ``None``. Function to use for converting
+    a sequence of string columns to an array of datetime instances. The default
+    uses dateutil.parser.parser to do the conversion. Pandas will try to call
```
backticks around ``dateutil.parser.parser`` *maybe* put in a refernce to it

frankcleary replied:
Fixed

jreback commented on doc/source/io.rst:
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
+    <https://docs.python.org/3/library/codecs.html#standard-encodings>`_.
+  - ``squeeze``: boolean, default ``False``. If the parsed data only contains
+    one column then return a Series.
+  - ``na_filter``: boolean, default ``True``. Detect missing value markers
```
let's put the NA options (``na_values``,``na_filter``,``keep_default_na``) near each other (change the code as well to match this ordering)

frankcleary replied:
Did the major reorg as suggested

jreback commented:
@frankcleary looks good. couple of points. 

I would like to have the options in the same order as ``io.rst`` sections. grouping if possible (e.g. the NA values together). I think this will be more logical.

All that said the more important options should be closer to the front if at all possible (I think that is how we had it). e.g. things like ``index_col``, ``header``, ``sep``. 

So ok to reorg a bit (see if you can keep the code in sync as well). everything is passed by kw so that should be fine. Add a note in the whatsnew to say that the ordering of args was changed (in API section) as well.

frankcleary commented:
Makes sense, thanks. I will incorporate all those suggestions next weekend and post an update by Monday Jan. 18.

frankcleary commented:
Here's the order I'm thinking of in terms of logically organizing the arguments, along with the general idea behind the grouping I put them in. Any thoughts?

##### Basic
filepath_or_buffer
sep
delimiter
##### Column and index names and location
header
names
index_col
usecols
squeeze
prefix
mangle_dupe_cols
##### General parsing details
dtype
engine
converters
true_values
false_values
skipinitialspace
skiprows
skipfooter
nrows
##### NA handling
na_values
keep_default_na
na_filter
verbose
skip_blank_lines
##### Date handling
parse_dates
infer_datetime_format
keep_date_col
date_parser
dayfirst
thousands
decimal
##### Iteration
iterator
chunksize
##### Quoting and file format
compression
lineterminator
quotechar
quoting
escapechar
comment
encoding
dialect
tupleize_cols
##### Errors
error_bad_lines
warn_bad_lines

jreback commented:
@frankcleary a reorg like that would great!

(note you can prob use the ``^^^^`` to make these sub-sections in the docs (I think)

frankcleary commented:
I've got an update almost ready with the reorg and other suggested changes, but need to fix merge conflicts in the parsers.py docstring so I won't be able to finish it tonight. I will get it done this weekend.

Is this about what you're thinking in regard to the sections:

<img width="787" alt="screen shot 2016-02-03 at 10 53 41 pm" src="https://cloud.githubusercontent.com/assets/5619798/12808069/26ff52ac-cac9-11e5-9bdc-f86ef78f244e.png">

jorisvandenbossche commented:
Thanks for this!

One consideration:
- I think we should try to make the actual option name stand out a bit more (not it is a bit lost in the dense bullet point). Maybe using something like http://docutils.sourceforge.net/docs/user/rst/quickref.html#definition-lists ? (just an idea, would have to try out to see how it looks)

frankcleary committed:
DOC: pd.read_csv doc-string clarification #11555

Updated IO Tools documentation for read_csv() and read_table() to be consistent with the doc-string
and reorded keyword arguements to group them more logically. Also updated concat docs in merging.rst
to be consistent with doc-string.

frankcleary commented:
Sorry, I was trying to push a clean history and it looks things are finished for this PR (can't reopen on a closed PR with a force pushed branch). I opened a new one here: https://github.com/pydata/pandas/pull/12256
